### DISCUSSION TITLE
Is it possible to fine-tune GitHub Copilot for custom repositories?

### DISCUSSION THREAD

**Discussion by sakashita-mk** [community/community#61307] [Copilot Conversations]
### Select Topic Area

Question

### Body

Hello everyone,
I have been using GitHub Copilot and I'm interested in fine-tuning it for my custom repositories. Is it possible to customize the capabilities of GitHub Copilot to align with specific domains or projects?

I would greatly appreciate any advice or information regarding the following questions:

1. Can GitHub Copilot be fine-tuned?
2. Are there any methods available to apply GitHub Copilot to custom repositories?
3. If there are any relevant documentation or guides, please provide information on how to access them.

Thank you in advance.

**Comment by Akash1134** [comment]
Hi @sakashita-mk .

Thanks for sharing this request. There is significant complexity involved in implementing something like this at the moment. 

The code suggested and accepted is viewed like any line of code a developer would add manually. This is why it is important to work on your code with the same principles and workflow than before GitHub Copilot:
- Write tests (GitHub Copilot can help)
- Automate as much as possible in CI
- Put the PR/Code review as the center of the collaboration

Domain specific functionality definitely something that would add a lot of value and we really appreciate your feedback. 

Thanks!

**Reply by sakashita-mk** [reply]
Hi @Akash1134 .

Thank you for your comment.
I understand that GitHub Copilot does not currently provide fine-tuning capabilities.
Do you have any plans to implement such functionality in the future? 

Thank you and best regards.

**Comment by elgalu** [comment]
Any plans to put this on the roadmap? For large enterprises with big code bases it becomes quite relevant that the model takes into account the proprietary source code, specially if they're already in the GH or GHE platform. Copilot competitors are jumping into this, e.g.

Related: https://codeium.com/blog/what-github-copilot-lacks-finetuning-on-your-private-code
> We prove that fine-tuning Codeium’s generic base code model on unseen code leads to substantial, observable improvements in suggestion quality over other tools such as GitHub Copilot

Related: https://stackoverflow.com/a/72844485/511069
> OpenAI API offers the "Davinci Codex" machine learning model with a pay-per-hit subscription, similar to the the non-coding version of the davinci model. OpenAI should enable the fine-tuning option to Davinci Codex as well. When they do it you will be able to use it via API calls.

**Comment by github-actions** [comment]
🕒 **Discussion Activity Reminder** 🕒

This Discussion has been labeled as dormant by an automated system for having no activity in the last 60 days. Please consider one the following actions:

1️⃣ Close as Out of Date: If the topic is no longer relevant, close the Discussion as `out of date` at the bottom of the page.

2️⃣ Provide More Information: Share additional details or context — or let the community know if you've found a solution on your own.

3️⃣ Mark a Reply as Answer: If your question has been answered by a reply, mark the most helpful reply as the solution.

Note: This dormant notification will only apply to Discussions with the `Question` label. To learn more, see our [recent announcement](https://github.com/orgs/community/discussions/70478).

Thank you for helping bring this Discussion to a resolution! 💬

**Comment by Akash1134** [comment]
[Custom models for GitHub Copilot are now in Limited Public Beta](https://github.blog/changelog/2024-08-27-custom-models-for-github-copilot-are-now-in-limited-public-beta/) ✅
